n3fit/src/n3fit/hyper_optimization/rewards.py

```diff
@@ -24,30 +24,300 @@
     std: 0.8925
 
 """
+import logging
+from typing import Callable, Dict, List
+
 import numpy as np
 
-from n3fit.vpinterface import N3PDF
+from n3fit.backends import MetaModel
+from n3fit.vpinterface import N3PDF, compute_phi2
+from validphys.core import DataGroupSpec
 from validphys.pdfgrids import distance_grids, xplotting_grid
 
+log = logging.getLogger(__name__)
+
 
 def _pdfs_to_n3pdfs(pdfs_per_fold):
     """Convert a list of multi-replica PDFs to a list of N3PDFs"""
     return [N3PDF(pdf.split_replicas(), name=f"fold_{k}") for k, pdf in enumerate(pdfs_per_fold)]
 
 
-def average(fold_losses=None, **_kwargs):
-    """Returns the average of fold losses"""
-    return np.average(fold_losses)
-
-
-def best_worst(fold_losses=None, **_kwargs):
-    """Returns the maximum loss of all k folds"""
-    return np.max(fold_losses)
-
+class HyperLoss:
+    """
+    Class to compute the hyper_loss based on the individual replica losses.
+
+    Computes the statistic over the replicas and then over the folds, both
+    statistics default to the average.
+
+    Parameters
+    ----------
+        loss_type: str
+            the type of loss over the replicas to use.
+            Options are "chi2" and "phi2".
+        replica_statistic: str
+            the statistic over the replicas to use, for per replica losses.
+            Options are "average", "best_worst", and "std".
+        fold_statistic: str
+            the statistic over the folds to use.
+            Options are "average", "best_worst", and "std".
+    """
 
-def std(fold_losses=None, **_kwargs):
-    """Return the standard dev of the losses of the folds"""
-    return np.std(fold_losses)
+    def __init__(
+        self, loss_type: str = None, replica_statistic: str = None, fold_statistic: str = None
+    ):
+        self.implemented_stats = {
+            "average": self._average,
+            "best_worst": self._best_worst,
+            "std": self._std,
+        }
+        self.implemented_losses = ["chi2", "phi2"]
+
+        self._default_statistic = "average"
+        self._default_loss = "chi2"
+
+        self.loss_type = self._parse_loss(loss_type)
+        self.reduce_over_replicas = self._parse_statistic(replica_statistic, "replica_statistic")
+        self.reduce_over_folds = self._parse_statistic(fold_statistic, "fold_statistic")
+
+        self.phi2_vector = []
+        self.chi2_matrix = []
+
+        self.penalties = {}
+
+    def compute_loss(
+        self,
+        penalties: Dict[str, np.ndarray],
```

Comment:
@RoyStegeman. They are essentially the same. The only thing is that they are converted to a dictionary here:
https://github.com/NNPDF/nnpdf/blob/d5b25b43e9eb9cdc190cb0ca769357a74f1ed5da/n3fit/src/n3fit/model_trainer.py#L996-L999
before entering `compute_loss`; this will later on make it easier handling these penalties in the overall statistics. Just in case, I updated the docstring and example of `compute_loss` to clarify it.